### PR TITLE
Add Binary Cross Entropy With Logit Loss to nn crate

### PR DESCRIPTION
**Pull request by ToluClassics**

### CODE REVIEW

File: candle-nn/src/loss.rs [hunk 1]
```diff
@@ -48,3 +48,27 @@ pub fn cross_entropy(inp: &Tensor, target: &Tensor) -> Result<Tensor> {
...
+///
+/// * [inp]: The input tensor of dimensions `N, C` where `N` is the batch size and `C` the number
+///          of categories. This is expected to raw logits.
+/// * [target]: The ground truth labels as a tensor of u32 of dimension `N, C` where `N` is the batch size and `C` the number
+///          of categories.
+///
+/// The resulting tensor is a scalar containing the average value over the batch.
+pub fn binary_cross_entropy_with_logit(inp: &Tensor, target: &Tensor) -> Result<Tensor> {
+    let inp = crate::ops::sigmoid(inp)?;
+    let one_tensor = Tensor::new(1.0, &inp.device())?;
```
Comment (LaurentMazare) [inline candle-nn/src/loss.rs:65]: This would only work for `f64` tensors and not for `f16`, `bf16`, `f32` and would allocate a tensor object. Instead you could use `target.affine(1., -1.)` etc which would avoid these drawbacks.

File: candle-nn/tests/loss.rs [hunk 2]
```diff
@@ -39,3 +39,38 @@ fn nll_and_cross_entropy() -> Result<()> {
     assert_eq!(to_vec0_round(&loss, 4)?, 1.1312);
     Ok(())
 }
+
```
Comment (LaurentMazare) [inline candle-nn/tests/loss.rs:42]: Could you put in as a comment the equivalent PyTorch snippet that can be used to check the test final output?

**Comment by LaurentMazare** [comment]
Oh also there seems to be some test failures in the CI that you want to check.

**Comment by ToluClassics** [comment]
> Oh also there seems to be some test failures in the CI that you want to check.

Yeah I notices, just fixed them in the new commit

**Comment by LaurentMazare** [comment]
Just to mention that the cuda CIs are expected to fail, no need to look into these!
